File: LongestSubstringWithoutRepeatingCharacters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        int start = 0, rez=0;
        // a b c a b c b
        //         ^
        // rez = 3-0=3
        // mapa: (a, 3) (b, 4) (c, 2)
        Dictionary<char, int> mapa = new Dictionary<char, int>();
        for(int i=0; i<s.Length; i++) {
            if(mapa.ContainsKey(s[i])) {
                int tmp = mapa[s[i]];
                foreach(KeyValuePair<char, int> kvp in mapa) {
                    if(kvp.Value <= tmp) {
                        mapa.Remove(kvp.Key);
                    }
                }
                start = Math.Max(start, tmp + 1);
            }
            rez = Math.Max(rez, i - start+1);
            mapa[s[i]] = i;
        }
        
        return rez;
    }
}

File: LongestSubstringWithoutRepeatingCharacters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> mapa = new HashMap<>();
        int longest = 0;
        int start = 0; // This will track the start index of the current substring

        for (int i = 0; i < s.length(); i++) {
            char currentChar = s.charAt(i);
            
            // If the character is already in the map and its index is within the current window
            if (mapa.containsKey(currentChar) && mapa.get(currentChar) >= start) {
                // Move the start index to one past the previous index of the current character
                start = mapa.get(currentChar) + 1;
            }
            
            // Update the map with the current character's index
            mapa.put(currentChar, i);
            
            // Update the length of the longest substring found so far
            longest = Math.max(longest, i - start + 1);
        }

        return longest;
    }
}
